Strings shouldn't support any divide-like binary op.

BrianPugh opened the pull request:
This allows for the correct implementation of `__rtruediv__` for classes like `pathlib.Path`.

See related discussion:

https://github.com/micropython/micropython-lib/issues/621

BrianPugh committed:
py/objstr.c: Strings shouldn't support any divide-like binary op.

This allows for the correct implementation of __rtruediv__ for classes
like pathlib.Path.

Signed-off-by: Brian Pugh <bnp117@gmail.com>

dlech commented:
Maybe better to just replace `bad_implicit_conversion(rhs_in);` with `return MP_OBJ_NULL;`?

BrianPugh commented:
> Maybe better to just replace `bad_implicit_conversion(rhs_in);` with `return MP_OBJ_NULL;`?

That'd be much cleaner 😄 . Let me change that.

BrianPugh committed:
py/objstr.c: Return MP_OBJ_NULL for incompatible rhs type.

Signed-off-by: Brian Pugh <bnp117@gmail.com>

BrianPugh commented:
implementing via just returning `MP_OBJ_NULL;` actually breaks some valid tests; will wait for more feedback before taking further action.

dpgeorge commented:
Does this most recent change allow implementing `__rtruediv__` as above?  Regardless, this PR would need to come with a test for this.

BrianPugh committed:
Revert "py/objstr.c: Return MP_OBJ_NULL for incompatible rhs type."

This reverts commit d7be1740b772195ee52f87b1c21f17e77c0c1ed0.

BrianPugh committed:
tests/basic/string_div.py: Initial commit.

Signed-off-by: Brian Pugh <bnp117@gmail.com>

BrianPugh commented:
@dpgeorge I reverted the commit to the previous, less clean (but working) solution. I also added a unit test. However, tests like:

```
val = ""
val /= DivClass()
print(val)  # rtruediv
```

still fail, and I don't really have a great environment setup to debug it.

dpgeorge commented:
See alternative at #11478, which gets the `val /= DivClass()` case working.

dpgeorge commented:
Closed in favour of the above PR, which was merged in ea7031faff9efe6803d8a8f67ad2e3b4a6d390e3
